### DISCUSSION TITLE
Treat terminal more like an editor

### DISCUSSION THREAD

**Discussion by rgbkrk** [zed-industries/zed#25482] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

When editing a shell command, I'd love to be able to click into the middle of the text to position my cursor.

Most terminal emulators don't support this, except for Warp and Kitty. I would _love_ for the feel of the terminal to match the seamless fluid feel of the editor.

The following is what I _think_ we need for this feature

1. Mouse-based cursor positioning
   - Allow users to click anywhere in the command line to position the cursor
   - Support selecting text with mouse drag

2. Full text editor capabilities for command input
   - Multi-line editing support
   - Copy/paste functionality
   - Multiple cursors and selections

3. Visual enhancements
   - Change cursor style from block to vertical bar (|)
   - Implement syntax highlighting for shell commands

4. Maintain shell compatibility
   - Ensure compatibility with popular shells (Bash, Zsh, Fish)
   - Implement hooks for shell integration (`precmd`, `preexec`)

### If applicable, add mockups / screenshots to help present your vision of the feature

Here's a totally fake way to show it

https://github.com/user-attachments/assets/1992de58-4baf-4350-b0b3-66c4abfff5a6

I'd _love_ if the cursor was a `|` too instead of a block.

**Comment by notpeter** [comment]
Nifty. I wonder exactly how Warp and Kitty have implemented this.

Native MacOS Terminal, Alacritty and others have some support for mouse events (via ANSI events, google `mouse "\e[?1003h"` for more).  A simple test to test:

```
brew install
dialog --yesno Testing 10 30
```

**Comment by rgbkrk** [comment]
There's two main articles from Warp on how they handle input as well as how terminals are generally implemented:

* [Why is the Terminal Input So Weird](https://www.warp.dev/blog/why-is-the-terminal-input-so-weird)
* [How Warp Works](https://www.warp.dev/blog/how-warp-works)

Based on this, I'll update the issue with what I _think_ we need.

**Comment by rgov** [comment]
Seems like this is close enough to my feature request: on macOS Terminal.app, you can position the cursor by option-clicking. I rely on this feature heavily. I'd like the behavior to match so it's not jarring switching between different apps. I think VS Code is also consistent here.

**Comment by notpeter** [comment]
I'm moving this to a discussion, but preserving the scoped request for `option-click` caret placement as:
- https://github.com/zed-industries/zed/issues/25481

**Comment by asqarslanov** [comment]
Maybe, instead of implementing a new terminal in-house, Zed can make use of existing implementations. The `libghostty` library looks promising.

**Comment by 42piratas** [comment]
https://github.com/zed-industries/zed/pull/58758
